---
title: "04_Plotting_TPM"
author: "JR"
date: "2024-08-04"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#install.packages("reshape")
library(reshape)
library(tidyr)
library(tibble)
library(ggplot2)
library(dplyr)
source("../../util/00_our_first_function_lesson.R")
```
# Goal: use TPM values to make plots of how genes change in abundance due to dox and time
So far we have been learning to plot using the results of DESEQ2 such as pvalue, padj 
Now we are going to take another step in plotting the abundance of each gene (TPM).
This would be the type of plot you do after finding sig genes in DESQ2 to see the trend of 
abundance and time. Let's use our figure skills and learn some more R and ggplot2 to measure gene expression levels.

# Important: Counts are a much more accurate representation of how much an RNA was observed in a sample
TPM compresses all those thousands of reads on each gene into one value (Transcripts per million).
TPM is just simply saying what percentage of a given sequencing sample a given gene is.
We also saw that almost all our data is at value TPM < 20 and std is very high -
So take all of this with a grain of salt but is good to measure relative changes in gene across samples!

# Let's start making some TPM plots !

# Load data
```{r loading our RNAseq analysis data TPM & COUNTS}

# load TPM RESULTS: 
load("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/TPM_results/TPM_results.Rdata")

# laod DESEQ2 results
load("/scratch/Shares/rinnclass/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/counts_results_DESEQ/DESEQ_results.rdata")

```

# First let's revisit the distribution of values for TPM 
```{r hist TPM values}
# Previously we plotted TPM distribution from our cols labeled:
# "WT_0_1" etc - note that is not a good idea to name a col by a number :)
hist(average_and_stddev_values$96_avg)
# Oh no this doesn't work because of an "integer" value being indexed
# This can be fixed easily by :
hist(average_and_stddev_values$'96_avg', xlim = c(0,1000), ylim = c(0,1000), breaks = 1000)
# We can quickly see again that most values are very low and a long right tail on distribution of TPM values!
```

# Now let's do this with Ggplot and learn to melt data :) Or pivot longer
Our goal here is to make a distribution of TPM values in each time point - using facet-wrap
```{r facet all timepoint TPM distributions}

# First let's make set of time values for faceting:
time_values <- c("0_avg", "12_avg", "24_avg", "48_avg", "96_avg")

# This can also be done with grep (smarter):
time_points <- names(average_and_stddev_values)[grep("avg", names(average_and_stddev_values))]
# moving rows into df so we can track gene_id
average_and_stddev_values_2 <- rownames_to_column(average_and_stddev_values, var = "gene_id")

# Melt the data frame to long format
melt_tpm_df <- melt(average_and_stddev_values_2, measure.vars = time_points, value_name = "gene_id")

# Cool we just made it a really long df (compare original and melt) -it's 5x longer
# Note the same can be done with pivot longer
pivot_longer_tpm_df <- average_and_stddev_values_2 %>% 
  pivot_longer(cols = all_of(time_points), names_to = 'Time', values_to = 'TPM')

# Now to plot:
ggplot(pivot_longer_tpm_df, aes(x = value)) +
  geom_histogram(bins = 50) +
  facet_wrap(~ variable, scales = "free_x")+
  xlim(0,1000) +
  ylim(0,1000) +
  theme("paperwhite")

# or

ggplot(pivot_longer_tpm_df, aes(x = TPM)) +
  geom_histogram(bins = 50) +
  facet_wrap(~ Time, scales = "free_x")+
  xlim(0,1000) +
  ylim(0,1000) +
  theme("paperwhite")

# Just for fun we can also get some of this in summary
summary(average_and_stddev_values)
```
# RESULT: average TPM distributions are similar in all samples


# Let's grab the counts just for the filtered res_df genes.
# Also let's make sure we have a list of unique genes!
```{r TPM for DESEQ2 sig genes}

sig_genes <- as.data.frame(unique(filtered_res_df$gene_id))

names(sig_genes) <- "gene_id"

# one way to subset the data frame of avrg / sd TPM values
sig_gene_tpm <- inner_join(average_and_stddev_values_2, sig_genes, by = "gene_id")

# Same can be done by indexing
sig_gene_tpm_2 <- average_and_stddev_values_2[average_and_stddev_values_2$gene_id %in%
                                                sig_genes$gene_id, ]

# let's now merge in the gene names into this handy dataframe
sig_gene_tpm <- merge(sig_gene_tpm, g2s)

# Now we should organize so we can plot specific genes (not mix character and integer)
sig_gene_tpm <- sig_gene_tpm %>%
  select(-gene_id) 

# cool now we have the tpm and sd values to plot !

```


# Now let's plot one gene TPM and sd across time, let's look in df for a good one (filtered_res_df)
```{r TPM plot of gene of interest}

# Picking a favorite gene
gene_up <- as.data.frame(c("Gm9045"))

names(gene_up) <- "gene_name"

# indexing to just get values for this gene
Gm9045_tpm <- sig_gene_tpm[sig_gene_tpm$gene_name %in% gene_up$gene_name, ]

# now moving the character gene_name to rownames to won't interfere with numeric values
rownames(Gm9045_tpm) <- Gm9045_tpm$gene_name

# removing text column
Gm9045_tpm <- Gm9045_tpm %>%
  select(-gene_name) 

# Ok this is typical data wrangling now let's plot !

```

# Line plot for TPM and SD across time points
Here we are going to leverage the fact that we already calculated
the mean and sd TPM values for all genes. So we will plot from that file
here. However, it is also useful to be able to take any gene from any table and do this.
We will do that next.
```{r plotting line plot}

# We now need to futher prepare our data :)
# Let's get a variable for col names to plot
avg_cols <- grep("_avg", names(Gm9045_tpm), value = TRUE)

sd_cols <- grep("_sd", names(Gm9045_tpm), value = TRUE)

# This is another way to "melt" or pivot longer
plot_df <- data.frame(
  time = as.numeric(gsub("_avg", "", avg_cols)),
  avg = unlist(Gm9045_tpm[avg_cols]),
  sd = unlist(Gm9045_tpm[sd_cols])
)

# NOW THE PLOT !
ggplot(plot_df, aes(x = time, y = avg)) +
  geom_line() +
  geom_errorbar(aes(ymin = avg - sd, ymax = avg + sd), width = 0.2) +
  theme("paperwhite") +
  labs(x = "Time (hours)", y = "TPM Value")
# +   theme_classic()

```

# Line plot from full TPM data frame.
Here we are going to do something more typical is plot the same line
plots above but using the replicate information. This way we are more 
flexible going to the full table with replicates. For example, we could 
calculate standard error instead of standard deviation this way.

# Logic flow:
(1) Have a set of genes of interest
(2) Reduce TPM table to those genes
(3) Pivot longer so all TPM data is in one column (genes have multiple rows)
(4) Group by and summarize to get mean and sd 
(5) Plot those values

```{r line plot from large table}

# (1) set of genes of interest or list.
genes_of_interest <- data.frame(gene = c("ENSMUSG00000094125.1", "ENSMUSG00000079247.2", "ENSMUSG00000112027.1", "ENSMUSG00000066632.3", "ENSMUSG00000000031.16"))

# (2) Reduce TPM file to these genes
df_plot_genes <- TPM_filtered[rownames(TPM_filtered) %in% genes_of_interest$gene, ]

# Moving rownames in to a column
df_plot_genes <- rownames_to_column(df_plot_genes, var = "gene_id")

# (3) Pivot longer (watch the changes !)
df_plot_genes <- df_plot_genes %>% 
  pivot_longer(cols = -gene_id, names_to = 'Time_Replicate', values_to = 'TPM')

# Seperating sample name into time and replicate for plotting table
df_plot_genes <- df_plot_genes %>%
  separate(
    Time_Replicate,
    into = c('WT', 'Time', 'Replicate'),
    sep = '_',
    extra = "merge"
  ) %>%
  select(-WT)

# Convert Time to numeric values
df_plot_genes$Time <- as.numeric(df_plot_genes$Time)

# (4) Group by Time and calculate mean and std of TPM
df_plot_genes <- df_plot_genes %>% 
  group_by(gene_id, Time) %>% 
  summarise(mean_TPM = mean(TPM), sd_TPM = sd(TPM))

# (5) Plot the line plot with error bars
ggplot(df_plot_genes, aes(x = Time, y = mean_TPM)) +
  geom_line() +
  geom_errorbar(aes(ymin = mean_TPM - sd_TPM, ymax = mean_TPM + sd_TPM), width = 0.2) +
  labs(x = 'Time (hours)', y = 'Average TPM') +
  ggtitle('Average TPM across Time Points') +
  facet_wrap(~ gene_id)

# This could be cleaned up by using gene_name instead -- fun exeRcise :)
```

########### exeRcise ################

# Goal (1) : can you make the same plot of lfC from DESEQ2 results?
Exact same logic, different starting data frame. This is a very
common plot to show LfC across time. Let's make it happen !

# Goal (2) : We should make a function !!! 
Wouldn't it be cool to have a funciton that we can pick any 
set of genes and make the plots we just did above ???
Yes we can - let's attempt to bundle that code for TPM file 
into a function !

#####################################

# Function
```{r exeRcise make our first function !}

# Alright let's make our first function !
#Start by calling a funciton to do all things between "{}" { function }
# We also just need to set the parameters for the funciton that need to be in env

# First name function, inputs and then the funciton items - kinda like shell script
plot_gene_expression <- function(genes_of_interest, df) {
  # Now we are going to do all the steps inside the function {}
  df_plot_genes <- df[rownames(df) %in% genes_of_interest, ]
  
  # Move rownames to a column
  df_plot_genes <- rownames_to_column(df_plot_genes, var = "gene_id")
  
  # Pivot longer
  df_plot_genes <- df_plot_genes %>%
    pivot_longer(cols = -gene_id,
                 names_to = 'Time_Replicate',
                 values_to = 'TPM')
  
  # Separate sample name into time and replicate
  df_plot_genes <- df_plot_genes %>%
    separate(
      Time_Replicate,
      into = c('WT', 'Time', 'Replicate'),
      sep = '_',
      extra = "merge"
    ) %>%
    select(-WT)
  
  # Convert Time to numeric values
  df_plot_genes$Time <- as.numeric(df_plot_genes$Time)
  
  # Group by Time and calculate mean and std of TPM
  df_plot_genes <- df_plot_genes %>%
    group_by(gene_id, Time) %>%
    summarise(mean_TPM = mean(TPM), sd_TPM = sd(TPM))
  
  # Plot the line plot with error bars
  ggplot(df_plot_genes, aes(x = Time, y = mean_TPM)) +
    geom_line() +
    geom_errorbar(aes(ymin = mean_TPM - sd_TPM, ymax = mean_TPM + sd_TPM), width = 0.2) +
    labs(x = 'Time (hours)', y = 'Average TPM') +
    ggtitle('Average TPM across Time Points') +
    facet_wrap( ~ gene_id)
}

# Example usage:
genes_of_interest <- c(
  "ENSMUSG00000094125.1",
  "ENSMUSG00000079247.2",
  "ENSMUSG00000112027.1",
  "ENSMUSG00000066632.3",
  "ENSMUSG00000000031.16"
)

plot_gene_expression(genes_of_interest, TPM_filtered)
```

# Cool now we can source this function from our util !!
Let's make an Rscript that has this function so we can use it like any other function !
1) Go to util folder
2) set as working directory !
3) make new .R script file
# Check out example util/00_our_first_function_lesson.R 
4) source below and function runs !

```{r source .R scripts }

source("../../util/00_our_first_function_lesson.R")

genes_of_interest <- c(
  "ENSMUSG00000094125.1",
  "ENSMUSG00000079247.2",
  "ENSMUSG00000112027.1",
  "ENSMUSG00000066632.3",
  "ENSMUSG00000000031.16"
)

plot_gene_expression(genes_of_interest, TPM_filtered)

```

# LFC plot 
Here we are going to follow a similar logic as above to intersect
a list of genes and plot the lfc on y-axis across time on x-axis
# Note we are using a little more succinct logic here!
```{r lfc versus time plot}

# Same as above starting with list of genes:
genes_of_interest <- data.frame(gene = c("ENSMUSG00000094125.1", "ENSMUSG00000079247.2", "ENSMUSG00000112027.1", "ENSMUSG00000066632.3", "ENSMUSG00000000031.16"))

# Reduce DESEQ2 results "res_df" to genes of interest
df_lfc_plot <- res_df[res_df$gene_id %in% genes_of_interest$gene, ]

# Remove the "time_point_" prefix and "_vs_0" suffix, then convert to numeric
df_lfc_plot <- df_lfc_plot %>%
  mutate(time_point = gsub("time_point_|_vs_0", "", result_name)) %>%
  mutate(time_point = as.numeric(time_point))

# Ensure the data is sorted by gene and time_point
df_lfc_plot <- df_lfc_plot %>%
  arrange(gene_id, time_point)

# Plotting the LFC versus time
ggplot(df_lfc_plot, aes(x = time_point, y = log2FoldChange, group = gene_id)) +
  geom_line() +
  geom_point() +
  facet_wrap(~ gene_id, scales = "free_y") +  # Facet by gene
  labs(title = "Log2 Fold Change Over Time",
       x = "Time Point (hours)",
       y = "log2 Fold Change") +
  theme_minimal()

```
# Awesome these plots look a little better too :) 
We don't have error bars here as each point only has one value
from DESEQ2. But it shows the trends a little better. Super handy !!

# Now let's make it a function !
```{r create LFC plot function }

LFCplot <- function(genes_of_interest, res_df) {
  library(dplyr)
  library(ggplot2)
  
  # Reduce DESEQ2 results to genes of interest
  df_lfc_plot <- res_df[res_df$gene_id %in% genes_of_interest$gene, ]
  
  # Remove the "time_point_" prefix and "_vs_0" suffix, then convert to numeric
  df_lfc_plot <- df_lfc_plot %>%
    mutate(time_point = gsub("time_point_|_vs_0", "", result_name)) %>%
    mutate(time_point = as.numeric(time_point))
  
  # Ensure the data is sorted by gene and time_point
  df_lfc_plot <- df_lfc_plot %>%
    arrange(gene_id, time_point)
  
  # Plotting the LFC versus time
  plot <- ggplot(df_lfc_plot, aes(x = time_point, y = log2FoldChange, group = gene_id)) +
    geom_line() +
    geom_point() +
    facet_wrap(~ gene_id, scales = "free_y") +  # Facet by gene
    labs(title = "Log2 Fold Change Over Time",
         x = "Time Point (hours)",
         y = "log2 Fold Change") +
    theme_minimal()
  
  return(plot)
}

```
# Now let's test it !
```{r}

source("../../util/00_our_first_function_lesson.R")

LFCplot(genes_of_interest, res_df)

```
# AMAZING !! one function to make any gene plot !
Have fun trying differnet gene inputs for TPM or LFC !



