---
title: "05_TPM_Transcripts_per_million"
author: "JR"
date: "2024-07-19"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
```


# Goal to do some analysis with TPM values.
We used the counts output from Salmon in the previous classes to do
differential expression and other analysis. All while starting to expand our R skills !

Now we are going to use Transcripts Per Million reads. These are normalized abundance
values of each gene. We can think of this as "how much" of a gene there is.
TPM will be our input for some plots and statistical tests as well.

# I couldn't explain it better or more simply than this site:
https://www.rna-seqblog.com/rpkm-fpkm-and-tpm-clearly-explained/

# Let's get started and load our TPM file from our NF_CORE RNAseq pipeline output
```{r reading in TPM}

# reading in table TPM
TPM <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/data_for_class/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

view(TPM)

#now we need to remove the gene name collumn

TPM <- TPM %>%
  select(-gene_name)

# Ok so we can see that we are missing our gene_names but have gene_id and a TPM value for each gene across each timepoint
# NOTE: rownames are gene_id

```

# Let's see how many genes are never expressed
```{r removing 0 value data}

# We are going to do this by indexing into TPM and grabbing all rows that add up to 0
TPM_zero <- TPM[rowSums(TPM == 0) == ncol(TPM), ]
view(TPM_zero)
# Wow so almost half of the genes measured had no expression

# now let's remove these from our TPM table 
# There are numerous ways to do this, here is one !

# first lets move the gene_id into a column to cross compare between TPM and TPM_zero
view(TPM)
TPM <- rownames_to_column(TPM, "gene_id")
view(TPM)
# same for TPM_zero
TPM_zero <- rownames_to_column(TPM_zero, "gene_id")

# Now we can remove these from TPM using the Tidyverse %in% 'function'
TPM_not_zero <- TPM[!(TPM$gene_id %in% TPM_zero$gene_id), ]
view(TPM_not_zero)

```
# Note:
if you filter by 0's as we did the starting number of genes is much higher.
This can matter in later statistical analysis when you have actually made more 
observations than your down stream analyses. This is important for some statistical principles.
Just remember you can always go back to the original TPM file we loaded to know how many observations 
were originally made.

# Let's start to get a feel for TPM data by seing what their values are like !
```{r Distibution of TPM values }

#first the distribution of TPM
hist(TPM_not_zero$WT_0_1, xlim = c(0,300), breaks = 1000)
# weird we still see a lot of small values (e.g., 0.2314). let's filter our TPM object differently

# First some cleaning up
# Moving gene_id column back to rownames
view(TPM)
row.names(TPM) <- TPM$gene_id
view(TPM)

# now removing the actual gene_id col
TPM$gene_id <- NULL
view(TPM)

# last clean up to move rownames back 
TPM <- column_to_rownames(TPM, "gene_id")
view(TPM)

# Let's filter to a row that adds up to at least 1
TPM_filtered <- TPM[rowSums(TPM) > 1, ]

# Let's check 
any(rowSums(TPM_filtered) < 1)

# For a positive control
any(rowSums(TPM) < 1)
# Cool look like we are filtered to any row that adds up to at leat 1 tpm

# Let's see if that histogram changed?
hist(TPM_filtered$WT_0_R1, xlim = c(0,300), breaks = 1000)

# Lets compare means 
mean(TPM_filtered$WT_0_R1)

# Now our non filtered mean
mean(TPM_not_zero$WT_0_R1)
```
# Result: we see that in one sample the average is about 30-45 TPM
That is for only one time point and replicate.
# Result: when we filtered on row sum > 1 TPM this mean shifts to 45 TPM


# Let's get to know TPM data a little better by looking across all samples
We can do this with rowMeans which is like row sum, but will get the mean value of a row

```{r average TPM across all samples}
# First let's take the aveargae across the row
row_means <- rowMeans(TPM_filtered)

# Look at the mean
mean(row_means)
# Cool so an average TPM value is about 45 across all samples
# Not just the one we time point we looked at above.

# What about the standard deviation?
sd(row_means)

# WOW that is a lot of standard deviation !

# Now let's look at the distribution of the average TPM for each gene.
hist(row_means, xlim = c(0,100), breaks = 1000)

```
# Result: We can see that the mean TPM is 45 but SD is 310 !
# Result: most are less than 20 though !
#Note: 
Any time you get a data set it's good to look at the distribution of values.
This will also help suggest statistical tests depending on the distribution type.
In short, always check the mean and SD as a first check on things !

# Now let's find out how many genes have a low TPM average
```{r how many genes have mean TPM of less than one}

# This is a nested set of code to mix rowMeans with indexing.
# what is the 'length' of mean row in TPM_filtered and then index for < 1
length(rowMeans(TPM_filtered)[rowMeans(TPM_filtered) < 1])
# Yikes about a quarter of our data has an average TPM vlaue of < 1
# We could think of filtering those out too

```
# Result: 8,333 genes have an average of TPM 1 across all samples.
This might seem confusing since we filtered for a row sum greater than one.
But a gene could be .1 TPM across 12 samples and the row would sum to greater than one.
Thus, we have 8K+ cases where expression is super low in all samples.

# Thought exercise how can I get the average TPM for each time point?
You have three columns for each timepoint and 5 time points. So we need a way 
to specify what we are looking for. This can be done by the function 'grep'. you
can think of it as find and replace. In summary we can find the columns for the timepoints
for each gene and calculate mean and standard deviaiton (sd) and so on ...

# Now let's see how to get the average value for a given time point for each gene
Let's use this as an opportunity to explore for loops more too :)

```{r for loop to find average TPM for each time point }

# First establish time point and replicate cols
time_points <- c("0", "12", "24", "48", "96")
replicates <- c("_1", "_2", "_3")

X_average_values <- list()
for (tp in time_points) {
  # Subset the columns for the current time point and replicates
  cols <- grep(paste0("WT_", tp, "_"), colnames(TPM_filtered))
  
  # Calculate the mean of the replicates for each row
  avg <- rowMeans(TPM_filtered[, cols])
  
  # Add the result to the list
  X_average_values <- c(X_average_values, list(avg))
}

# Convert the list to a data frame
X_average_values <- do.call(cbind, X_average_values)

# Add column names for the time points
colnames(X_average_values) <- time_points

# convert back to data.frame
avrg_TPM <- as.data.frame(X_average_values)

# Let's take a look
view(avrg_TPM)
```
# Nice we have a dataframe with the average TPM for each timepoint for each gene!
# Note:
This can also be done with the function 'lapply' which is essentially a forloop maker.
We will use this later -- just to note 'lapply' is commonly used instead of forloops (faster too)


# What about standard deviation (sd)? This is critical information when considering abundance measurements.
Let's use the same logic as above and do this for both sd and mean TPM for all genes in TPM filtered

```{r for loop to find average TPM value and sd for each gene at each time point}

# To get the standard deviation we need this package.
install.packages("matrixStats")
library(matrixStats)


# Let's establish our time points and replicates there in:
# First establish time point and replicate cols
time_points <- c("0", "12", "24", "48", "96")
replicates <- c("_1", "_2", "_3")

?rowSds
# Initialize empty lists to store the results from the for loop
average_and_stddev_values <- list()

# Loop through each time point
# now each object into the for loop is "tp"
for (tp in time_points) {
  # Subset the columns for the current time point and replicates
  cols <- grep(paste0("WT_", tp, "_"), colnames(TPM_filtered))
  
  # Calculate the mean and standard deviation of the replicates for each row
  avg <- rowMeans(TPM_filtered[, cols])
  std_dev <- apply(TPM_filtered[, cols], 1, sd)
  
  # Convert std_dev to a data frame
  std_dev <- data.frame(std_dev)
  
  # Combine avg and std_dev into one data frame
  combined <- cbind(avg, std_dev)
  
  # Add the result to the list
  average_and_stddev_values <- c(average_and_stddev_values, list(combined))
}

# Convert the list to a data frame
average_and_stddev_values <- do.call(cbind, average_and_stddev_values)

# Add column names for the time points
colnames(average_and_stddev_values) <- paste0(rep(time_points, each = 2), c("_avg", "_sd"))

# Voila
view(average_and_stddev_values)
```
# Nice we have a table with average and sd for TPM values we can use for Plotting and making figures !
# >>>> next we will combine all our previous lessons succinctly for an RNAseq analysis pipeline (workflow)

############### ExeRcise ##################

# 1) Double check that the averages and sd are correct by comaping TPM and average_and_standard_deviation .....
# 2) can you do the average and standard deviation using lapply? If so please share !

