---
title: "DESEQ_results_script"
author: "JR"
date: "2024-07-23"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(IRanges)
library(dplyr)
library(tidyr)
library(tibble)
library(readr)
library(ggplot2)
library(purrr)
library(magrittr)
library(pheatmap)
library(textshape)
library(Rcpp)
library(DESeq2)
```

So far, we have done a lot of fundamental RNA sequencing analysis - great work!
Now let's compile this and compress it into one RMD file. This will be a great way to
(i) practice what we have learned, (ii) a useful resource for any future RNAseq you may want to do
and (iii) ALL the data we need to start making figures - our next section :)

# Goal: script our RNAseq analysis pipeline to find genes changing significant across time from dox exposure. 

# Importing counts and TPM values
```{r import counts and TPM from Salmon}
# laoding Salmon counts from NF_CORE run 3.14
counts_matrix <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/data_for_class/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

# preparing counts for DESEQ
# create g2s
g2s <- data.frame(
  gene_id = rownames(counts_matrix),
  gene_name = counts_matrix[, 1]
)
# removing gene_name column for analyses
counts_matrix <- counts_matrix[ , -1]
# Round counts to integer mode required for DESEQ2
counts_integer <- round(counts_matrix)

```

# Creating sample sheet for deseq (coldata) from counts file
```{r creating deseq sample sheet (coldata)}
# setting up data frame with cols from counts matrix
deseq_samples <- data.frame(
  sample_id = colnames(counts_matrix))
# creating time point and replicate cols from col names
# Split names
split_values <- strsplit(deseq_samples$sample_id, "_")
# graabbing time point and replicate info 
time_values <- sapply(split_values, function(x) x[[2]])
replicate_values <- sapply(split_values, function(x) x[[3]])
# placing time point and replicate info into dataframe
deseq_samples$time_point <- time_values
deseq_samples$replicate <- replicate_values
# IMPORTANT : Factor time point and replicate
deseq_samples$time_point <- factor(deseq_samples$time_point)
deseq_samples$replicate <- factor(deseq_samples$replicate)

```

# Running DESEQ2
```{r running deseq2}
# testing sample sheet and counts are arranged properly 
stopifnot(all(colnames(counts_integer) == rownames(deseq_samples$sample_id)))

# setting up dds
dds_time_point <- DESeqDataSetFromMatrix(countData = counts_integer,
                              colData = deseq_samples,
                              design = ~ time_point)
# Run time point model
dds_time_point <- DESeq(dds_time_point)

```

# Extracting results for each time point from DESEQ dds
```{r compile DESEQ2 results}

# setting up result names object for input to forloop
result_names <- resultsNames(dds_time_point)
results_names <- result_names[-1]

# Setting up dataframe
res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# run for loop to retreive results from all time comparisons in dds
for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(dds_time_point, name = results_name)
  tmp_res_df <- res %>% 
    as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  res_df <- bind_rows(res_df, tmp_res_df)
  
}

```

# Filtering to significant results P <0.05 and LFC abs(1)
```{r filtering DESEQ results to LFC > abs(1)& P <0.05}

filtered_res_df <- res_df %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)

unique_sig_genes <- unique(filtered_res_df$gene_id)
```

# Nice only 100 lines of code to go from counts table to DESEQ results !
Now your future self can go from raw sequencing files to NF_CORE pipeline to DESEQ analysis.
That should be pretty handy if you ever do RNAseq in the future.

# NOW LET'S SAVE ALL OUR WORK FOR MAKING FIGURES !
Since all this data is from counts, let's make a results directory
in our results directory :) called 'count_results_DESEQ'

```{r saving DESEQ results}

save(unique_sig_genes, counts_integer, g2s, res_df, filtered_res_df, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/DESEQ_results.rdata")

load("results/DESEQ_results.rdata")
```

# As a final step let's archive the average and standard deviation of our TPM values
After this last step we will have all the info we need for plotting in lv 7 !

# Import TPM file from Salmon
```{r import counts and TPM from Salmon}
# laoding and preparing TPM file from Salmon

TPM <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/data_for_class/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

TPM <- TPM %>%
  select(-gene_name)
# Filtering for TPM of at least 1 across samples.
TPM_filtered <- TPM[rowSums(TPM) > 1, ]
# Checking
any(rowSums(TPM_filtered) < 1)
any(rowSums(TPM) < 1)

```

# Calculating mean and sd for each gene in each time point (replicate mean and sd)
```{r making table of mean and sd TPM value for each gene for each time point}
#install.packages("matrixStats")
library(matrixStats)

# Time and replicate values
time_points <- c("0", "12", "24", "48", "96")
replicates <- c("_1", "_2", "_3")

# initialize list for results
average_and_stddev_values <- list()
# now each object into the for loop is "tp"
for (tp in time_points) {
  cols <- grep(paste0("WT_", tp, "_"), colnames(TPM_filtered))
  avg <- rowMeans(TPM_filtered[, cols])
  std_dev <- apply(TPM_filtered[, cols], 1, sd)
  std_dev <- data.frame(std_dev)
  combined <- cbind(avg, std_dev)
  average_and_stddev_values <- c(average_and_stddev_values, list(combined))
}
# Convert the list to a data frame
average_and_stddev_values <- do.call(cbind, average_and_stddev_values)
# Add column names for the time points
colnames(average_and_stddev_values) <- paste0(rep(time_points, each = 2), c("_avg", "_sd"))
```

# Now let's save our TPM results as a seperate object
```{r saving TPM data}

# note I am adding in some other files we just saved, this is in case
# we just want to open up TPM data we still have G2s and Samplesheet.
save(TPM, average_and_stddev_values, TPM_filtered, g2s, deseq_samples, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/TPM_results.Rdata" )

# test
load("results/TPM_results.Rdata")
```
# VERY NICE we have all our results organized and can start to make figures
# WE ARE READY FOR LEVEL 7 - Congrats on finishing Level 6 !!